main.c

```diff
@@ -331,12 +380,75 @@ multithreaded(struct task_t *task, struct config_t *config)
     return context.found;
 }
 
+void *
+gn_worker(void *arg)
+{
+    struct gn_context_t *context = (struct gn_context_t *) arg;
+    struct config_t *config = context->config;
+
+    struct st_context_t st_context;
+    st_context.hash = context->hash;
+    st_context.cd.initialized = 0;
+
+    while (true)
+    {
+        struct task_t task;
+        pthread_mutex_lock(&context->mutex);
+        task = *context->iter_state.task;
+        bool has_next = iter_next(&context->iter_state);
```

Comment:
iter_next gives false for the last generated password, but after that it will give once again true and repeat whole set of generated values. 
I.e. all threads will exit only after multiple iterations over whole set of generated values.